src/sw_mh_helper.cpp

```diff
@@ -868,6 +701,142 @@ int mh_decision(double mh_prob)
 
 }
 
+// Add constraints to match SMC and MS
+
+/*
+ * Helper function to iterate over constraints and apply them
+ */
+double add_constraint(const std::string& name, List constraints,
+                      IntegerVector districts, NumericVector &psi_vec,
+                      std::function<double(List, int)> fn_constr) {
+    if (!constraints.containsElementNamed(name.c_str())) return 0;
+
+    List constr = constraints[name];
+    double val = 0;
+    double psi = 0.0;
+    for (int i = 0; i < constr.size(); i++) {
+        List constr_inst = constr[i];
+        double strength = constr_inst["strength"];
+        if (strength != 0) {
+            for (int dist = 0; dist < districts.size(); dist++) {
+                psi = fn_constr(constr_inst, districts(dist));
+                psi_vec[name] = psi + psi_vec[name];
+                val += strength * (psi);
+            }
+        }
+    }
+    return val;
+}
+
+/*
+ * Add specific constraint weights & return the cumulative weight vector
+ */
+double calc_gibbs_tgt(const subview_col<uword> &plan, int n_distr, int V,
```

Comment:
is this the same as the one in M-S?  If so (or close enough that we could make it one function with a few if statements), we should merge them for sure.

Reply:
A `mcmc_gibbs.cpp` file included by both algorithms would do the trick, for instance

Reply:
It's pretty close. The main differences are:
- Flip can adjust many districts, so it takes a vector of districts to iterate over. That means MS can be written into that form, but it may slow it down a teeny bit?
- Flip tracks psi values, with this function

I think we can merge them, but let me know your thoughts.

Reply:
I'm not wedded to one way or the other since I don't touch the flip code but reducing redundant code seems like a priority for us & especially with this PR

Reply:
This is implemented now. Would appreciate a look especially at the MS part of it, as there is a vector introduced that doesn't do anything, but makes everything work.